### PR TITLE
Foreign Key Fix

### PR DESCRIPTION
**Pull request by vittvolt**
Fix the current foreign key handling logic. 

This PR shall support the following foreign key creation syntax:

`create table tb (int num references t(id) on delete cascade)`

The actions supported will be at least CASCADE, RESTRICT and NOACTION.
Match type is not supported.

### CODE REVIEW

**Comment by apavlo** [comment]
@vittvolt Let me know when you want someone to review this.

**Comment by apavlo** [comment]
@vittvolt Awesome! We actually may not want to worry about updating the old executor code. We could get somebody to implement in the LLVM engine.

**Comment by apavlo** [comment]
@vittvolt What's the status of this? Do we just need to resolve the conflicts and then review it?

**Comment by vittvolt** [comment]
@apavlo This is ready to be reviewed and merged. I'll rebase it first.

**Comment by apavlo** [comment]
@pervazea Can you take care of this and get it merged?

**Comment by pervazea** [comment]
@vittvolt You are in the process of rebasing, correct? If so, please change status to ready to review when rebase is complete.

**Comment by vittvolt** [comment]
@pervazea Yes I had a few issues after rebasing on my local branch. I'll push the changes after my flight.

File: src/concurrency/timestamp_ordering_transaction_manager.cpp [hunk 1]
```diff
@@ -637,8 +637,8 @@ void TimestampOrderingTransactionManager::PerformDelete(
 
   auto transaction_id = current_txn->GetTransactionId();
 
-  PL_ASSERT(GetLastReaderCommitId(tile_group_header, old_location.offset) ==
-            current_txn->GetCommitId());
+  // PL_ASSERT(GetLastReaderCommitId(tile_group_header, old_location.offset) ==
```
Comment (vittvolt) [inline src/concurrency/timestamp_ordering_transaction_manager.cpp]: This assertion does not pass when modifying another table in the executor.

**Comment by vittvolt** [comment]
@pervazea Ready for review.

**Comment by pervazea** [comment]
Starting to review. Will provide review comments shortly.

File: src/concurrency/timestamp_ordering_transaction_manager.cpp (hunk 1 above)
Comment (pervazea) [reply src/concurrency/timestamp_ordering_transaction_manager.cpp]: Ok. We shouldn't leave a commented out assertion here though, without an explanation. Do we understand if

- The assertion is wrong?
- The assertion is correct, but there is a bug, in the new code or existing code?

One or more of the following are suggested:

- Add a comment explaining why the assertion is commented out
- File an issue about it
- Understand what is going on, and fix or remove the assertion accordingly.

File: src/executor/create_executor.cpp [hunk 2]
```diff
@@ -139,7 +140,7 @@ bool CreateExecutor::CreateTable(const planner::CreatePlan &node) {
         // Sink Column Offsets
         std::vector<oid_t> sink_col_ids;
         for (auto col_name : fk.foreign_key_sinks) {
-          oid_t col_id = source_table->GetSchema()->GetColumnID(col_name);
+          oid_t col_id = sink_table->GetSchema()->GetColumnID(col_name);
```
Comment (pervazea) [inline src/executor/create_executor.cpp]: Source -> sink table change, is a bug fix?

File: src/parser/postgresparser.cpp [hunk 3]
```diff
@@ -894,26 +894,44 @@ parser::ColumnDefinition *PostgresParser::ColumnDefTransform(ColumnDef *root) {
       else if (constraint->contype == CONSTR_UNIQUE)
         result->unique = true;
       else if (constraint->contype == CONSTR_FOREIGN) {
-        result->table_info_.reset(new TableInfo());
         // Transform foreign key attributes
         // Reference table
-        result->table_info_->table_name = constraint->pktable->relname;
-        // Reference column
+
+        // additionally, add a speci·al ColumnDefinition to the list!
```
Comment (pervazea) [inline src/parser/postgresparser.cpp]: extraneous char, should be "special"

File: src/parser/postgresparser.cpp [hunk 4]
```diff
@@ -894,26 +894,44 @@ parser::ColumnDefinition *PostgresParser::ColumnDefTransform(ColumnDef *root) {
...
+          auto attr_cell = constraint->pk_attrs->head;
+          value* attr_val =
+                reinterpret_cast<value*>(attr_cell->data.ptr_value);
+          col->foreign_key_sink.emplace_back(attr_val->val.str);
+
+          LOG_ERROR("PK attr: %s\n", attr_val->val.str);
+        }
+        else
+        {
+          // TODO: Default to the primary key of the sink table
```
Comment (pervazea) [inline src/parser/postgresparser.cpp]: TODO: what code should be here?

File: src/parser/postgresparser.cpp [hunk 5]
```diff
@@ -894,26 +894,44 @@ parser::ColumnDefinition *PostgresParser::ColumnDefTransform(ColumnDef *root) {
...
             CharToActionType(constraint->fk_del_action);
-        result->foreign_key_update_action =
+        col->foreign_key_update_action =
             CharToActionType(constraint->fk_upd_action);
         // Match type
-        result->foreign_key_match_type =
-            CharToMatchType(constraint->fk_matchtype);
+        col->foreign_key_match_type = CharToMatchType(constraint->fk_matchtype);
+
+        LOG_ERROR("PK table name: %s.\n", col->fk_sink_table_name.c_str());
```
Comment (pervazea) [inline src/parser/postgresparser.cpp]: On an initial read.. LOG_ERROR statements report information, but what is the error they are trying to report? Similarly for line 911.

File: src/parser/postgresparser.cpp [hunk 6]
```diff
@@ -960,14 +978,14 @@ parser::SQLStatement *PostgresParser::CreateTransform(CreateStmt *root) {
     Node *node = reinterpret_cast<Node *>(cell->data.ptr_value);
     if ((node->type) == T_ColumnDef) {
       // Transform Regular Column
-      ColumnDefinition *temp = nullptr;
+      // ColumnDefinition* temp = nullptr;
```
Comment (pervazea) [inline src/parser/postgresparser.cpp]: Remove or explain why commented out.

File: src/parser/postgresparser.cpp [hunk 7]
```diff
@@ -960,14 +978,14 @@ parser::SQLStatement *PostgresParser::CreateTransform(CreateStmt *root) {
...
       try {
-        temp = ColumnDefTransform(reinterpret_cast<ColumnDef *>(node));
+        ColumnDefTransform(reinterpret_cast<ColumnDef*>(node), result);
       } catch (NotImplementedException e) {
         delete result;
         throw e;
       }
-      result->columns.push_back(std::unique_ptr<ColumnDefinition>(temp));
+      // result->columns.push_back(std::unique_ptr<ColumnDefinition>(temp));
```
Comment (pervazea) [inline src/parser/postgresparser.cpp]: Remove or explain why commented out.

File: src/parser/postgresparser.cpp [hunk 8]
```diff
@@ -978,8 +996,10 @@ parser::SQLStatement *PostgresParser::CreateTransform(CreateStmt *root) {
               reinterpret_cast<value *>(key_cell->data.ptr_value)->val.str);
         }
       } else if (constraint->contype == CONSTR_FOREIGN) {
+        // TODO: this routine will crash the program
```
Comment (pervazea) [inline src/parser/postgresparser.cpp]: Explain what needs to be done here.

File: src/planner/create_plan.cpp [hunk 9]
```diff
@@ -48,14 +48,19 @@ CreatePlan::CreatePlan(parser::CreateStatement *parse_tree)
...
       for (auto &col : parse_tree->columns) {
-        // The parser puts the Foreign Key information into an artificial
-        // ColumnDefinition.
-        if (col->type == parser::ColumnDefinition::DataType::FOREIGN) {
-          this->ProcessForeignKeyConstraint(table_name, col.get());
-          // XXX: Why should we always continue here?
-          continue;
-        }
+        
+        // if (col->type == parser::ColumnDefinition::DataType::FOREIGN) {
```
Comment (pervazea) [inline src/planner/create_plan.cpp]: These commented out lines. Not needed? Remove?

File: src/planner/create_plan.cpp [hunk 10]
```diff
@@ -214,7 +219,7 @@ void CreatePlan::ProcessForeignKeyConstraint(const std::string &table_name,
   fkey_info.constraint_name =
       "FK_" + table_name + "->" + fkey_info.sink_table_name;
 
-  LOG_DEBUG("Added a foreign key constraint toward sink table %s",
+  LOG_ERROR("Added a foreign key constraint toward sink table %s",
```
Comment (pervazea) [inline src/planner/create_plan.cpp]: LOG_DEBUG -> LOG_ERROR? If an error, what is the error?

File: src/storage/data_table.cpp [hunk 11]
```diff
@@ -593,13 +775,13 @@ bool DataTable::CheckForeignKeyConstraints(
...
-        std::vector<oid_t> key_attrs = foreign_key->GetSourceColumnIds();
+        std::vector<oid_t> key_attrs = foreign_key->GetSinkColumnIds();
         std::unique_ptr<catalog::Schema> foreign_key_schema(
-            catalog::Schema::CopySchema(schema, key_attrs));
+            catalog::Schema::CopySchema(ref_table->schema, key_attrs));
         std::unique_ptr<storage::Tuple> key(
             new storage::Tuple(foreign_key_schema.get(), true));
         // FIXME: what is the 3rd arg should be?
-        key->SetFromTuple(tuple, key_attrs, index->GetPool());
+        key->SetFromTuple(tuple, foreign_key->GetSourceColumnIds(), index->GetPool());
```
Comment (pervazea) [inline src/storage/data_table.cpp]: I know 3rd argument is unchanged... re: the FIXME comment, is there still a question about the 3rd argument?

File: test/sql/foreign_key_sql_test.cpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  int rows_affected;
+
+  EXPECT_EQ(TestingSQLUtil::ExecuteSQLQuery(
+      "UPDATE tb1 SET id = 10 where id = 1;"), ResultType::SUCCESS);
+
+  TestingSQLUtil::ExecuteSQLQuery("SELECT id from tb1;", result,
+                                  tuple_descriptor, rows_affected,
+                                  error_message);
+
+  EXPECT_EQ("10", result[0]);
```
Comment (pervazea) [inline test/sql/foreign_key_sql_test.cpp]: In other tests, Prashanth recommends use of ExecuteSQLQueryAndCheckResult. Applicable in this test?

File: src/storage/data_table.cpp [hunk 13]
```diff
@@ -557,6 +569,175 @@ bool DataTable::InsertInSecondaryIndexes(const AbstractTuple *tuple,
   return res;
 }
 
+/**
```
Comment (pervazea) [inline src/storage/data_table.cpp]: Nice to see function documentation.

File: src/executor/create_executor.cpp (hunk 2 above)
Comment (vittvolt) [reply src/executor/create_executor.cpp]: This is a bug fix

**Comment by vittvolt** [comment]
@pervazea All issues addressed.

**Comment by apavlo** [comment]
@pervazea Let me know if it's okay to merge this.

@vittvolt Thanks so much for everything!

File: src/storage/data_table.cpp [hunk 14]
```diff
@@ -608,6 +789,22 @@ bool DataTable::CheckForeignKeyConstraints(
 
         // if this key doesn't exist in the refered column
         if (location_ptrs.size() == 0) {
+          LOG_DEBUG("The key does not exist in the sink table!\n");
```
Comment (pervazea) [inline src/storage/data_table.cpp]: I would log which key and which sink table.

File: src/storage/data_table.cpp [hunk 15]
```diff
@@ -608,6 +789,22 @@ bool DataTable::CheckForeignKeyConstraints(
...
+        auto tile_group_header = tile_group->GetHeader();
+
+        auto &transaction_manager =
+          concurrency::TransactionManagerFactory::GetInstance();
+        auto visibility = transaction_manager.IsVisible(
+          transaction, tile_group_header, location_ptrs[0]->offset,
+          VisibilityIdType::READ_ID);
+
+        if (visibility != VisibilityType::OK) {
+          LOG_DEBUG("The tuple is not visible yet.\n");
```
Comment (pervazea) [inline src/storage/data_table.cpp]: Is there any useful information about the tuple that should be logged here?

File: src/parser/postgresparser.cpp [hunk 16]
```diff
@@ -894,26 +895,41 @@ parser::ColumnDefinition *PostgresParser::ColumnDefTransform(ColumnDef *root) {
...
-                reinterpret_cast<value *>(attr_cell->data.ptr_value);
-            result->foreign_key_sink.push_back(std::string(attr_val->val.str));
-          }
+        {
+          auto attr_cell = constraint->pk_attrs->head;
+          value* attr_val =
+                reinterpret_cast<value*>(attr_cell->data.ptr_value);
+          col->foreign_key_sink.emplace_back(attr_val->val.str);
+
+          LOG_ERROR("PK attr: %s\n", attr_val->val.str);
```
[the remark this answers is not in the document]
Comment (vittvolt) [reply src/parser/postgresparser.cpp]: This was used as a debugging message. I don't think it is needed anymore.

**Comment by vittvolt** [comment]
@pervazea All checks passed, ready for merge!

**Comment by pmenon** [comment]
@pervazea Can you take a second pass? I noticed some unnecessary `UNUSED_ATTRIBUTE` macros and formatting issues.

**Comment by pervazea** [comment]
Sure.

There was also some aspects of the logging I wanted to clean up.

________________________________

**Pull request merged**